### DISCUSSION TITLE
Subdivision level for the emissive exoplanet meshes

### DISCUSSION THREAD

**Discussion by EricNelson-sketch** [CelestiaProject/Celestia#1882] [General]
When I go to the planets from Gurren Lagann's addons such as Gliese 486 b, some of the GAIA planets, and XI Aquilae b (Fortitudo), the planets use meshes, which's neat.

However, the subdivision level for the models is anything but excellent.

The quality isn't really bad, and not to be critical, but the subdivision level doesn't seem fully satisfying for bodies as large as planets, not to mention some planets are gas giants.

A subdivision level of 1 has 20 triangles and is extremely bad for planetary modelling
A subdivision level of 2 has 80 triangles and is very bad for planetary modelling.
A subdivision level of 3 has 320 triangles and is bad for planetary modelling.
A subdivision level of 4 has 1280 triangles and is not so good for planetary modelling.
A subdivision level of 5 has 5120 triangles and is below average for planetary modelling.
A subdivision level of 6 has 20480 triangles and is average for planetary modelling.
A subdivision level of 7 has 81920 triangles and is good for planetary modelling.
A subdivision level of 8 has 327680 triangles and is excellent for planetary modelling.

It looks like the planet meshes (which are by Cubic) are level 6, which's admirable from space, but not from aircraft cruising altitude or especially surface view.
Not to mention that's only 160 sides of an outline of the mesh.
Not to mention each level above 2 doubles the number of sides making the outline.
A subdivision level of 1 makes 6 or 10 sides of an outline depending on the viewing angle.
A subdivision level of 2 makes 10 sides of an outline.
A subdivision level of 3 makes 20 sides of an outline.
A subdivision level of 4 makes 40 sides of an outline.
A subdivision level of 5 makes 80 sides of an outline.
A subdivision level of 6 makes 160 sides of an outline.
A subdivision level of 7 equals 320 sides of an outline.
A subdivision level of 8 makes 640 sides of an outline.

A circle has 360 degrees, which's 160 x 2.25.
Which means each segment of the outline of every mesh of his takes up 2.25 degrees of their perimeter.
Each segment of a 6 sided outline takes up 60 degrees of the perimeter.
Each segment of a 10 sided outline takes up 36 degrees of the perimeter.
Each segment of a 20 sided outline takes up 18 degrees of the perimeter.
Each segment of a 40 sided outline takes up 9 degrees of the perimeter.
Each segment of an 80 sided outline takes up 4.5 degrees of the perimeter.
Each segment of a 160 sided outline takes 2.25 degrees of the perimeter.
Each segment of a 320 sided outline takes up 1.125 degrees of the perimeter.
Each segment of a 640 sided outline takes up 0.5625 degrees of the perimeter.

I admire the effort Cubic puts in for modeling the planets (in Gurren Lagann's addons), though they can be better.
Wondering what he uses to model those planets in cmod format if anything.
Though it's cool how these meshes help with the normal mapping and night texture all at once.
Though I know the only user named Gurren Lagann on this site is @GurrenLagannTSS

**Comment by SevenSpheres** [comment]
This is about addons, not Celestia's codebase, so it doesn't belong here.
The meshes in question are spherical; see [this thread](https://celestiaproject.space/forum/viewtopic.php?f=6&t=20578).

**Reply by EricNelson-sketch** [reply]
Approximately spherical you mean.
A sphere is a 3D figure where every set of points is equally distant from the center.
A polyhedron is a 3-dimensional figure with 4 or more faces/triangles.
Though we can accept the nature of what CG technology gives us.
Though I was referring to the code of the meshes (on these addons), not Celestia's code, in case you got confused.

**Reply by SevenSpheres** [reply]
As far as I can tell cubic's mesh is almost as spherical as Celestia's default built-in sphere mesh, which is... pretty close to a perfect sphere. Maybe it's slightly off near the poles but it's barely noticeable, especially from space.

> Though I was referring to the code of the meshes (on these addons), not Celestia's code, in case you got confused.

Which is why Celestia's GitHub repository was the wrong place to post this.

**Reply by EricNelson-sketch** [reply]
Almost is perceptual, though in reality, the mathematical difference is significant.
The closer to the default bodies in Celestia, the higher the subdivision level, and I realize that a subdivision level that's 2 levels greater than cubic's meshes is considered the magic value for approximating a great quality sphere by anyone's knowledge.
Besides, the closer to such bodies, the higher the subdivision level is, by a big difference, meaning the level of standing on the surface is several levels greater than viewing them at the distance where the bodies take up the height of a PC monitor, which on the other hand Cubic's levels compare to.
Besides, the pentagonal patterns at the poles are pretty distinct (even from space) for their size regarding cubic's meshes.
With 2 more subdivisions, the polar pentagons are a quarter as large as they are on Cubic's meshes, indicating a much better approximation to spherical, even from a human standing view.

**Comment by GurrenLagannTSS** [comment]
> Though I know the only user named Gurren Lagann on this site is @GurrenLagannTSS

That's indeed the late-teen that you are looking for.
For how was it done, see Seven's reply above.

**Reply by EricNelson-sketch** [reply]
Yeah, Cubic worked on the mesh, though if I used his tools, I could've made 2 more levels for the mesh than Cubic did, knowing that it would look great even from the surface.
Cubic's quality though is good when viewed far enough to see the body as a whole.

**Comment by 375gnu** [comment]
closing as unrelated. please use the forum for this topic.

**Reply by EricNelson-sketch** [reply]
As said.
